### PR TITLE
add autonat v2 spec

### PR DESCRIPTION
**Pull request by sukunrt**
This is the same as #538. That PR targeted `autonat-rename` branch which has a commit to move the autonat v1 spec to a separate v1 specific file. 

Note: When merging, make sure to use "Rebase and Merge" to ensure git blame works correctly for autonat v1 spec.

### CODE REVIEW

File: autonat/autonat-v2.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+3. `autonat v2` provides a mechanism for nodes to dial an IP address different
+from the requesting node's observed IP address without risking amplification
+attacks. `autonat v1` disallowed such dials to prevent amplification attacks.
+
+
+## AutoNAT V2 Protocol
+
+![Autonat V2 Interaction](autonat-v2.svg)
```
Comment (MarcoPolo) [inline autonat/autonat-v2.md:63]: These diagrams need to be updated. They use “DialStatuses” and reference a undefined error.

File: autonat/autonat-v2.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+`E_DIAL_REFUSED`: The server didn't dial back any address because it was
+incapable of dialing or unwilling to dial any of the requested addresses.
+
+`E_INTERNAL_ERROR`: Error not classified within the above error codes occured on
+server preventing it from completing the request.
+
+`OK`: the server completed the request successfully. A request is considered
+completed successfully when the server either completes a dial(successfully or
+unsuccessfully) or rejects all addresses in the request as undialable.
```
Comment (MarcoPolo) [inline autonat/autonat-v2.md:153]: If a server rejects all addresses as undialable, should it set OK or E_DIAL_REFUSED?

File: autonat/autonat-v2.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+### Amplification Attack Prevention
+
+![Interaction](autonat-v2-amplification-attack-prevention.svg)
+
+When a client asks a server to dial an address that is not the client's observed
+IP address, the server asks the client to send some non trivial amount of bytes
+as a cost to dial a different IP address. To make amplification attacks
+unattractive, servers SHOULD ask for 30k to 100k bytes. Since most handshakes
+cost less than 10k bytes in bandwidth, 30kB is sufficient to make attacks
```
Comment (MarcoPolo) [inline autonat/autonat-v2.md:165]: Linking to QUIC’s similar mitigation here would be helpful

File: autonat/autonat-v2.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+are free to use different heuristics than this one
+
+Servers SHOULD NOT reuse their listening port when making a dial back. In case
+the client has reused their listen port when dialing out to the server, not
+reusing the listen port for attempts prevents accidental hole punches. Clients
+SHOULD only rely on the nonce and not on the peerID for verifying the dial back
+as the server is free to use a separate peerID for the dial backs.
+
+Servers SHOULD determine whether they have IPv6 and IPv4 connectivity. IPv4 only servers SHOULD refuse requests for dialing IPv6 addresses and IPv6 only
+servers SHOULD refuse requests for dialing IPv4 addresses.
```
Comment (MarcoPolo) [inline autonat/autonat-v2.md:217]: Do we currently handle this in go-libp2p?

**Review by MarcoPolo — APPROVED** [review]

File: autonat/autonat-v2.md (hunk 4 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:217]: Yes. We use the black hole detector to tell us if we have IPv6 connectivity. https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/autonatv2/server.go#L183

File: autonat/autonat-v2.md (hunk 2 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:153]: good catch! it should be E_DIAL_REFUSED

File: autonat/autonat-v2.md (hunk 1 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:63]: done.

File: autonat/autonat-v2.md (hunk 3 above)
Comment (sukunrt) [reply autonat/autonat-v2.md:164]: added some text.

File: autonat/autonat-v2-amplification-attack-prevention.plantuml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+participant Cli
+participant Srv
+
+skinparam sequenceMessageAlign center
+skinparam defaultFontName monospaced
+
+
+== Amplification Attack Prevention ==
+
+Cli -> Srv: [conn1: stream: dial] DialRequest:{nonce: 0xabcd, addrs: (addr1, addr2, addr3)}
```
Comment (MarcoPolo) [inline autonat/autonat-v2-amplification-attack-prevention.plantuml:11]: Very small nit, but can we 0-index these addrs too? `addr0, addr1, addr2`

**Review by MarcoPolo — APPROVED** [review]
I pushed a couple small changes as well fyi

**Review by sukunrt — COMMENTED** [review]
:man_shrugging:

**Pull request merged**